apps/volunteer/sign_up.py

```diff
@@ -81,6 +79,47 @@ def update_volunteer_from_form(volunteer, form):
 @volunteer.route("/sign-up", methods=["GET", "POST"])
 @feature_flag("VOLUNTEERS_SIGNUP")
 def sign_up():
+
+    # Work out our first and last arrivals based on config
+    first_arrival = event_start() - timedelta(app.config["ARRIVAL_DAYS"])
+    last_departure = event_end() + timedelta(app.config["DEPARTURE_DAYS"])
+
+    # Add the first date to the list with or earlier
+    ARRIVAL_CHOICES.append((
```

Comment:
This should only run once, not every time this endpoint is called (otherwise every time someone signs up the arrival choices list will expand).

I think this is why there's the `generate_day_options` function.

Reply:
Good catch, I completely missed that in testing.

New commit checks the length of the dates lists to see if they're empty before attempting to populate them.

Reply:
It could be preferable to assign these values at the same time as they're declared in the module; i.e. something along the lines of:

```
# Declare and initialize arrival and departure lists
ARRIVAL_CHOICES = generate_arrival_choices()
DEPARTURE_CHOICES = generate_departure_choices()
```

That'd ensure that the variables are only calculated once, and makes it easier to add automated test coverage, since the tests won't need to invoke `sign_up` before checking the arrival & departure choices.

This could also help other developers to read and understand the module.  The fact that the variables are uppercase implies that they may be [constants according to Python's style guide](https://www.python.org/dev/peps/pep-0008/#constants).  Another developer who is aware of the uppercase variable convention might assume (incorrectly) that `ARRIVAL_CHOICES = []`, for example, implies that the variable would always contain an empty list.

Reply:
@jayaddison you missed the discussion where we worked out that wasn't possible because it uses config from the app object which isn't available at module scope. 

It has to run within a view function to use the config. At any rate it's definitely not a performance issue.

Reply:
Ah, sorry - my mistake; ignore my previous comment.  I'll read up on the discussion before adding anything further.

Reply:
Findings: `wtforms` recommend implementing dynamically-calculated choices by assigning values to the [choices list](https://github.com/emfcamp/Website/blob/2d5cae652076f7aa8c2defac92ba102ba60a1e6c/apps/volunteer/sign_up.py#L38-L41) (more info: [code example](https://github.com/wtforms/wtforms/issues/185#issuecomment-396902599), [docs](https://wtforms.readthedocs.io/en/stable/fields.html#wtforms.fields.SelectField)).

The docs also mention that the `choices` setting is only evaluated once -- so if we're going to use the `ARRIVAL_CHOICES, DEPARTURE_CHOICES` variables, it'll be important to make sure they're assigned before requests reach the endpoint.

For robustness it'd be nice to prove that valid signup dates are accepted and that invalid ones are rejected (perhaps by copying & modifying an [existing endpoint test](https://github.com/emfcamp/Website/blob/2d5cae652076f7aa8c2defac92ba102ba60a1e6c/tests/test_users_app.py#L25-L39) and modifying it for `/volunteer/sign_up`?).

Reply:
Thanks for the pointer to the dynamic choices docs, hopefully the updated version is a better implementation.